fdbclient/NativeAPI.actor.cpp

```diff
@@ -221,6 +221,37 @@ void DatabaseContext::addSSIdTagMapping(const UID& uid, const Tag& tag) {
 	ssidTagMapping[uid] = tag;
 }
 
+void DatabaseContext::getLatestCommitVersion(const StorageServerInterface& ssi,
+                                             Version readVersion,
+                                             VersionVector& latestCommitVersion) {
+	latestCommitVersion.clear();
+
+	if (ssVersionVectorCache.getMaxVersion() == invalidVersion) {
+		return;
+	}
+
+	// Error checking (based on the assumption that the read version was not obtained
+	// from the client's grv cache).
+	if (readVersion > ssVersionVectorCache.getMaxVersion()) {
+		TraceEvent(SevError, "GetLatestCommitVersion")
+		    .detail("ReadVersion", readVersion)
+		    .detail("VersionVector", ssVersionVectorCache.toString());
+		ASSERT(false);
+	}
```

Comment:
There is code duplication here (with "getLatestCommitVersions()"), but refactoring while providing all the information needed to log the "CommitVersionNotFoundForSS" trace event (in the other method) might need a little bit of work. Let me know what you think. Thanks!

Reply:
if you made the loop in lines 243-252 a separate function called `setVersionVectorForTag()` it may make the code more readable for both `getLatestCommitVersion(s)`

Reply:
Note that "getLatestCommitVersion()" inserts the commitVersion in "latestCommitVersion" whereas "getLatestCommitVersions()" simply inserts the commitVersion in "versionMap", so I don't think we can factor out the "set" part. Maybe we can factor out the "get" part, and let the receiver insert the returned commitVersion (if any) in the appropriate map? Also, the "CommitVersionNotFoundForSS" trace event (in "getLatestCommitVersions()") needs to know whether the tag was found in the "ssidTagMapping" or not, so we would need to return that information too.